pipeline/api/asyncio/cloud.py

```diff
@@ -0,0 +1,242 @@
+from pipeline.schemas.file import FileCreate, FileGet
+from pipeline.schemas.function import FunctionGet
+from pipeline.schemas.model import ModelGet
+from pipeline.schemas.pipeline import PipelineFileVariableGet, PipelineGet
+from pipeline.schemas.pipeline_file import MultipartUploadMetadata, PipelineFileGet
+from pipeline.schemas.run import RunCreate, RunGet
+from pipeline.util import generate_id, python_object_to_hex
+
+if TYPE_CHECKING:
+    from pipeline.objects import Function, Graph, Model
+
+from pipeline.api import PipelineCloud as _SyncPipelineCloud
+
+FILE_CHUNK_SIZE = 200 * 1024 * 1024  # 200 MiB
+
+
+class PipelineCloud(_SyncPipelineCloud):
+    def _raise_not_implemeneted(self):
+        raise NotImplementedError("This function is not implemented")
+
+    def upload_python_object_to_file(self, obj, remote_path) -> FileGet:
+        self._raise_not_implemeneted()
+
+    def _initialise_direct_pipeline_file_upload(self, file_size: int) -> str:
+        self._raise_not_implemeneted()
+
+    def _direct_upload_pipeline_file_chunk(
+        self, data: bytes, pipeline_file_id: str, part_num: int
+    ) -> MultipartUploadMetadata:
+        self._raise_not_implemeneted()
+
+    def _finalise_direct_pipeline_file_upload(
+        self, pipeline_file_id: str, multipart_metadata: List[MultipartUploadMetadata]
+    ) -> PipelineFileGet:
+        self._raise_not_implemeneted()
+
+    def upload_pipeline_file(self, pipeline_file) -> PipelineFileVariableGet:
+        self._raise_not_implemeneted()
+
+    def _get(self, endpoint: str, params: Dict[str, Any] = None):
+        self._raise_not_implemeneted()
+
+    def upload_function(self, function: Function) -> FunctionGet:
+        self._raise_not_implemeneted()
+
+    def upload_model(self, model: Model) -> ModelGet:
+        self._raise_not_implemeneted()
+
+    def upload_pipeline(
+        self,
+        new_pipeline_graph: Graph,
+        public: bool = False,
+        description: str = "",
+        tags: Set[str] = None,
+    ) -> PipelineGet:
+        self._raise_not_implemeneted()
+
+    def _download_schema(
+        self, schema: Type[BaseModel], endpoint: str, params: Optional[Dict[str, Any]]
+    ) -> Type[BaseModel]:
+        self._raise_not_implemeneted()
+
+    def download_function(self, id: str) -> Function:
+        self._raise_not_implemeneted()
+
+    def download_model(self, id: str) -> Model:
+        self._raise_not_implemeneted()
+
+    def download_data(self, id: str) -> Any:
+        self._raise_not_implemeneted()
+
+    def download_result(self, result_id_or_schema: Union[str, RunGet]) -> Any:
+        self._raise_not_implemeneted()
+
+    def download_pipeline(self, id: str) -> Graph:
+        self._raise_not_implemeneted()
+
+    @staticmethod
+    def _get_raise_for_status(response: httpx.Response) -> None:
+        # A handler for errors that might be sent with messages from Top.
+        if not response.status_code == httpx.codes.OK:
```

Comment:
there's still quite a bit of code duplication across the async and sync client, for example this method is identical to the sync version apart from this one line. Maybe it's worth making this one line work with both `httpx` and `requests`? Or we could just switch to `httpx` entirely since it supports both sync and async?

@alexpearce - any opinions on this?

Reply:
I've added in something to fix this using a union typing input and checking if the response is httpx or requests. In a follow up PR we can consolidate the library used, no need at the moment I don't think.